File: modules/ontrack-voice-lists/src/OnTrackVoiceListsModule.ts
import { NativeModule, requireOptionalNativeModule } from 'expo';

import type { VoicePendingOp } from './OnTrackVoiceLists.types';

declare class OnTrackVoiceListsModule extends NativeModule<{
  onPending(): void;
}> {
  publishSnapshotAsync(json: string): Promise<void>;
  takePendingAsync(): Promise<VoicePendingOp[]>;
}

export default requireOptionalNativeModule<OnTrackVoiceListsModule>(
  'OnTrackVoiceLists',
);

File: modules/ontrack-voice-lists/android/src/main/java/expo/modules/ontrackvoicelists/OnTrackVoiceListsModule.kt
package expo.modules.ontrackvoicelists

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class OnTrackVoiceListsModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("OnTrackVoiceLists")
    Events("onPending")

    OnCreate {
      OnTrackVoiceStore.onPending = {
        sendEvent("onPending", emptyMap<String, Any>())
      }
    }

    OnDestroy {
      OnTrackVoiceStore.onPending = null
    }

    AsyncFunction("publishSnapshotAsync") { json: String ->
      val context = appContext.reactContext ?: appContext.currentActivity
        ?: error("The app is not ready.")
      OnTrackVoiceStore.publishSnapshot(context.applicationContext, json)
    }

    AsyncFunction("takePendingAsync") {
      val context = appContext.reactContext ?: appContext.currentActivity
        ?: error("The app is not ready.")
      OnTrackVoiceStore.takePending(context.applicationContext)
    }
  }
}
